hs_core/tasks.py

```diff
@@ -443,6 +443,61 @@ def copy_resource_task(ori_res_id, new_res_id=None, request_username=None):
         raise utils.ResourceCopyException(str(ex))
 
 
+@shared_task
+def replicate_resource_bag_to_user_zone_task(res_id, request_username):
+    """
+    Task for replicating resource bag which will be created on demand if not existent already to iRODS user zone
+    Args:
+        res_id: the resource id with its bag to be replicated to iRODS user zone
+        request_username: the requesting user's username to whose user zone space the bag is copied to
+
+    Returns:
+    None, but exceptions will be raised if there is an issue with iRODS operation
+    """
+
+    res = utils.get_resource_by_shortkey(res_id)
+    res_coll = res.root_path
+    istorage = res.get_irods_storage()
+    bag_modified_flag = True
+    # needs to check whether res_id collection exists before getting/setting AVU on it to
+    # accommodate the case where the very same resource gets deleted by another request when
+    # it is getting downloaded
+    if istorage.exists(res_coll):
+        bag_modified = istorage.getAVU(res_coll, 'bag_modified')
+
+        # make sure bag_modified_flag is set to False only if bag exists and bag_modified AVU
+        # is False; otherwise, bag_modified_flag will take the default True value so that the
+        # bag will be created or recreated
+        if bag_modified:
+            if bag_modified.lower() == "false":
+                bag_file_name = res_id + '.zip'
+                if res.resource_federation_path:
+                    bag_full_path = os.path.join(res.resource_federation_path, 'bags',
+                                                 bag_file_name)
+                else:
+                    bag_full_path = os.path.join('bags', bag_file_name)
+
+                if istorage.exists(bag_full_path):
+                    bag_modified_flag = False
+
+        if bag_modified_flag:
+            # import here to avoid circular import issue
+            create_bag_by_irods(res_id)
+
+        # do replication of the resource bag to irods user zone
+        if not res.resource_federation_path:
+            istorage.set_fed_zone_session()
+        src_file = res.bag_path
+        tgt_file = '/{userzone}/home/{username}/{resid}.zip'.format(
+            userzone=settings.HS_USER_IRODS_ZONE, username=request_username, resid=res_id)
+        fsize = istorage.size(src_file)
+        utils.validate_user_quota(request_username, fsize)
+        istorage.copyFiles(src_file, tgt_file)
+        return res.get_absolute_url()
+    else:
```

Comment:
Wondering if Exception should be caught to append a message that includes the resource id before raising again.

Reply:
The iRODS exception includes resource id, so I thought I'll just let whatever exception being raised from called methods to propagate rather than catching them explicitly here.